### PR TITLE
[AIR] Add tuner test.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?
1. Add tuner tests.
These tests are mainly focusing on non ray client mode, including successful runs, and failures in both driver and trainer side and resume.
2. One issue surfaced through writing the tests (which probably means the API is not quite right) is whether RunConfig should be supplied in Tuner.__init__ v.s. Tuner.fit(). At least for some fields in RunConfig, we want to be able to change it across runs (e.g. callbacks). Plus with current impl, it's not possible to checkpoint "stateful" callbacks, which could confuse our users. cc @ericl for API inputs. See "test_tuner_with_xgboost_trainer_driver_fail_and_resume" (search for hack).
3. The PR also cleans up some API docs.
4. Fixes some bugs in loading trial from checkpoint, namely `get_default_resource` (which probably is not necessary given self.placement_group_factory is already set anyways) is called with an empty `config`, as `self.config` is only loaded through `__setstate__`, which happens later than `get_default_resource`. Remove the call to `get_default_resource` when loading trials from checkpoint.

## Related issue number

#23328

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tuner.py [hunk 1]
```diff
@@ -164,7 +164,7 @@ def fit(self) -> ResultGrid:
                 raise TuneError(
                     f"Tune run failed."
                     f'Please use tuner = Tuner.restore("'
-                    f'{self._local_tuner.experiment_checkpoint_dir}") to resume.'
+                    f'{self._local_tuner.experiment_checkpoint_dir()}") to resume.'
```
Comment (krfricke) [inline python/ray/tune/tuner.py:167]: Not directly related, but if @property doesn't work, can we use `get_experiment_checkpoint_dir` instead as a name? Otherwise it's confusing as we have the `experiment_checkpoint_dir` property (not as a function) in several other locations

**Review by krfricke — APPROVED** [review]
Test looks good!

File: python/ray/tune/tuner.py (hunk 1 above)
Comment (xwjiang2010) [reply python/ray/tune/tuner.py:167]: That's a good point!
Changed to `get_experiment_checkpoint_dir` instead.

**Review by xwjiang2010 — COMMENTED** [review]
@krfricke Sorry, this probably needs another pass as quite some changes happened since.

**Comment by ericl** [comment]
I see, so the issue is that restoring from a checkpoint means the
constructor isn't called, and hence there's no opportunity to change the
Run config?

That makes sense as an argument to move it to fit() then.

File: python/ray/ml/config.py [hunk 2]
```diff
@@ -125,18 +125,22 @@ class RunConfig:
     This includes both running a Trainable by itself or running a hyperparameter
     tuning job on top of a Trainable (applies to each trial).
 
+    At resume, tune will automatically apply the same run config so that resumed
```
Comment (krfricke) [inline python/ray/ml/config.py:128]:
```suggestion
    At resume, Ray Tune will automatically apply the same run config so that resumed
```

File: python/ray/ml/config.py [hunk 3]
```diff
@@ -125,18 +125,22 @@ class RunConfig:
...
         name: Name of the trial or experiment. If not provided, will be deduced
             from the Trainable.
         local_dir: Local dir to save training results to.
             Defaults to ``~/ray_results``.
         callbacks: Callbacks to invoke.
             Refer to ray.tune.callback.Callback for more info.
+            Callbacks should be serializable.
+            Currently only supporting stateless Callback for resumed runs.
+            (Any state of the callback will not be checkpointed by Tune
+            and thus will not take effect in resumed runs).
```
Comment (krfricke) [inline python/ray/ml/config.py:141]:
```suggestion
            Currently only stateless callbacks are supported for resumed runs.
            (any state of the callback will not be checkpointed by Tune
            and thus will not take effect in resumed runs).
```

File: python/ray/tune/impl/tuner_internal.py [hunk 4]
```diff
@@ -151,7 +150,7 @@ def fit(self) -> ResultGrid:
             analysis = self._fit_internal(trainable, param_space)
         else:
             analysis = self._fit_resume(trainable)
-        analysis._legacy_checkpoint = False
+
```
Comment (krfricke) [inline python/ray/tune/impl/tuner_internal.py:153]: Why is this removed - shouldn't Tuner just return the new `Checkpoints`?

File: python/ray/tune/trial.py [hunk 5]
```diff
@@ -280,8 +287,9 @@ def __init__(
         # Parameters that Tune varies across searches.
         self.evaluated_params = evaluated_params or {}
         self.experiment_tag = experiment_tag
+        self.location = Location()
```
Comment (krfricke) [inline python/ray/tune/trial.py:290]: Duplicated in line 313 (remove there?)

**Review by krfricke — CHANGES_REQUESTED** [review]
Looks good, minor questions/nits

File: python/ray/tune/impl/tuner_internal.py (hunk 4 above)
Comment (xwjiang2010) [reply python/ray/tune/impl/tuner_internal.py:153]: It's a bit confusing, as ResultGrid is not using `ExperimentAnalysis.get_best_checkpoint` in any way, where `_legacy_checkpoint` is taking effect through. 

It's currently only using `ExperimentAnalysis.best_trial` and then `trial.checkpoint.value` --> convert to a new Checkpoint class.

Comment (krfricke) [reply python/ray/tune/impl/tuner_internal.py:153]: Ok, so to be clear, this means we are still always returning a `Checkpoint` and not a `TrialCheckpoint`, correct? Can we assert that in the test? `assert isinstance(cp, Checkpoint) and not isinstance(cp, TrialCheckpoint)`

Comment (xwjiang2010) [reply python/ray/tune/impl/tuner_internal.py:153]: Yes, that's correct. Added the assertion!

Comment (krfricke) [reply python/ray/tune/impl/tuner_internal.py:153]: I think we have to specifically check `not isinstance(cp, TrialCheckpoint)` as `TrialCheckpoint` is also a `Checkpoint`

**Review by krfricke — APPROVED** [review]
Awesome!
